fcrepo4/fcrepo4
# Ensures a valid id  and isPartOf in Message after Memento, and ACL creation, update and deletion

**Ensures a valid id  and non-null isPartOf in Message after Memento and ACL creation, update(acl) and deletion.***

**JIRA Ticket**:  https://jira.duraspace.org/browse/FCREPO-2872 and  https://jira.duraspace.org/browse/FCREPO-2874, and  https://jira.duraspace.org/browse/FCREPO-2873
* Other Relevant Links (Mailing list discussion, related pull requests, etc.)
 

# What does this Pull Request do?
As it says above, ensures that the @id in the JSON-LD message returns the URL to the affected resource and makes sure that a base URL is passed to Memento and ACL messages.
There is an open question in my mind about isPartOf:  currently the isPartOf value is always the repository root.  While the resources are part of the root I'm not sure how useful the field is.   Should the value of isPartOf depend on the object?  Should it reflect "containment" when containment applies? OriginalResource in the case of mementos? the "parent" of the ACL? 

# How should this be tested?
Setup a listener: https://wiki.duraspace.org/display/FEDORA4x/How+to+Inspect+Event+Messages+Generated+by+Fedora

Create a versioned resource: 
`curl -X POST -H "Link: <http://mementoweb.org/ns#OriginalResource>; rel=\"type\"" http://localhost:8080/rest/ -H "slug: testxyz" -u fedoraAdmin:fedoraAdmin`
Create a memento: 
`curl -X POST http://localhost:8080/rest/testxyz/fcr:versions -u fedoraAdmin:fedoraAdmin`
Verify that the message id reflects the newly created memento and the isPartOf is the root URL
Delete the memento:
`curl -X DELETE http://localhost:8080/rest/testxyz/fcr:versions/<memento-timestamp> -u fedoraAdmin:fedoraAdmin`
Verify that the message id reflects the deleted memento and the isPartOf is the root URL

Create a file with with following contents:  acl.ttl

```
@prefix acl: <http://www.w3.org/ns/auth/acl#> .

<#auth> a acl:Authorization ;
              acl:agent "testuser";
              acl:mode acl:Read;
              acl:accessTo </rest/testxyz> .

```
Create ACL
`curl -v -u fedoraAdmin:fedoraAdmin -X PUT --data-binary  "@acl.ttl" -H "Content-Type: text/turtle" http://localhost:8080/rest/testxyz/fcr:acl`

Verify that the message id reflects the newly created acl and the isPartOf is the root URL
Delete the acl:
`curl -X DELETE http://localhost:8080/rest/testxyz/fcr:acl -u fedoraAdmin:fedoraAdmin`
Verify that the message id reflects the deleted ACL and the isPartOf is the root URL


Example:
* Does this change require documentation to be updated? No
* Does this change/ add any new dependencies? adds fcrepo-http-commons to fcrepo-event-serialization
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)? no
* Could this change impact execution of existing code? no

# Interested parties
Tag (@ mention) interested parties or, if unsure, @fcrepo4/committers
